packages/components/src/composite/stories/legacy.story.tsx

```diff
@@ -0,0 +1,244 @@
+/**
+ * External dependencies
+ */
+import type { Meta, StoryFn, StoryContext } from '@storybook/react';
+
+/**
+ * WordPress dependencies
+ */
+import { isRTL } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import {
+	Composite,
+	CompositeItem,
+	CompositeGroup,
+	useCompositeState,
+} from '../legacy';
+
+import type { InitialState } from '../legacy';
+
+const Placeholder = ( _: InitialState ) => <></>;
+
+const transform = ( code: string, context: StoryContext ) => {
+	const config = ` ${ JSON.stringify( context.args, null, 2 ) } `;
+	const state = config.replace( ' {} ', '' );
+	return [
+		`const state = useCompositeState(${ state });`,
+		'',
+		'return (',
+		'  ' +
+			code
+				.replaceAll( /state=\{\{[\s\S]*?\}\}/g, 'state={ state }' )
+				.replaceAll( '=>', '' )
+				.replaceAll( /baseId=[^>]+?(\s*>)/g, ( _, close ) => {
+					return `{ ...state }${ close }`;
+				} )
+				.replaceAll( /<Composite\w+[\s\S]*?>/g, ( match ) =>
+					match.replaceAll( /\s+\s/g, ' ' )
+				)
+				.replaceAll(
+					/ >\s+([\w\s]*?)\s+<\//g,
+					( _, value ) => `>${ value }</`
+				)
+				.replaceAll( '} >', '}>' )
+				.replaceAll( /\n/g, '\n  ' ),
+		');',
+	].join( '\n' );
+};
```

Comment:
This transform function looks a bit over-engineered —I believe that [`paramenters.docs.source.type = 'code'`](https://storybook.js.org/docs/api/doc-block-source#type)  should be enough — and in case it isn't, I'd almost argue that we could write the source text directly for every story.

Reply:
Yes, `paramenters.docs.source.type = 'code'` is technically enough to render something, but it could be more helpful; it conveys the idea, but doesn't fully utilise the capabilities of Storybook. The same is true of writing the source directly, just with more duplication.

Personally I'm in favour of using `auto`/`dynamic` where possible, so that changes made to arg values in the data table are reflected in the code sample. Unfortunately, in this particular situation using `auto`/`dynamic` is equally unhelpful, because of what's automatically generated...

```jsx
<Composite
  aria-label="Legacy Composite with state prop (two dimensions)"
  state={{
    baseId: ':r66:',
    store: {
      __unstableInternals: {
        batch: () => {},
        init: () => {},
        omit: () => {},
        pick: () => {},
        setup: () => {},
        subscribe: () => {},
        sync: () => {}
      },
      __unstablePrivateStore: {
        __unstableInternals: {
          batch: () => {},
          init: () => {},
          omit: () => {},
          pick: () => {},
          setup: () => {},
          subscribe: () => {},
          sync: () => {}
        },
        getState: () => {},
        setState: () => {}
      },
      down: () => {},
      first: () => {},
      getState: () => {},
      item: () => {},
      last: () => {},
      move: () => {},
      next: () => {},
      previous: () => {},
      registerItem: () => {},
      renderItem: () => {},
      setActiveId: () => {},
      setBaseElement: () => {},
      setState: () => {},
      up: () => {},
      useState: () => {}
    }
  }}
>
  <CompositeGroup
    role="row"
    state={{
      baseId: ':r66:',
      store: {
        __unstableInternals: {
          batch: () => {},
          init: () => {},
          omit: () => {},
          pick: () => {},
          setup: () => {},
          subscribe: () => {},
          sync: () => {}
        },
        __unstablePrivateStore: {
          __unstableInternals: {
            batch: () => {},
            init: () => {},
            omit: () => {},
            pick: () => {},
            setup: () => {},
            subscribe: () => {},
            sync: () => {}
          },
          getState: () => {},
          setState: () => {}
        },
        down: () => {},
        first: () => {},
        getState: () => {},
        item: () => {},
        last: () => {},
        move: () => {},
        next: () => {},
        previous: () => {},
        registerItem: () => {},
        renderItem: () => {},
        setActiveId: () => {},
        setBaseElement: () => {},
        setState: () => {},
        up: () => {},
        useState: () => {}
      }
    }}
  >
    <CompositeItem
      state={{
        baseId: ':r66:',
        store: {
          __unstableInternals: {
            batch: () => {},
            init: () => {},
            omit: () => {},
            pick: () => {},
            setup: () => {},
            subscribe: () => {},
            sync: () => {}
          },
          __unstablePrivateStore: {
            __unstableInternals: {
              batch: () => {},
              init: () => {},
              omit: () => {},
              pick: () => {},
              setup: () => {},
              subscribe: () => {},
              sync: () => {}
            },
            getState: () => {},
            setState: () => {}
          },
          down: () => {},
          first: () => {},
          getState: () => {},
          item: () => {},
          last: () => {},
          move: () => {},
          next: () => {},
          previous: () => {},
          registerItem: () => {},
          renderItem: () => {},
          setActiveId: () => {},
          setBaseElement: () => {},
          setState: () => {},
          up: () => {},
          useState: () => {}
        }
      }}
    >
      Item A1
    </CompositeItem>
    // ...and so on for the other items...
  </CompositeGroup>
  // ...and the other groups.
</Composite>
```

With `transform` in place, that instead looks like this...

```jsx
// This is updated live as the arg table changes...
const state = useCompositeState( {
  "loop": true
} );

return (
  <Composite
    aria-label="Legacy Composite with state prop (two dimensions)"
    state={ state }
  >
    <CompositeGroup role="row" state={ state }>
      <CompositeItem state={ state }>Item A1</CompositeItem>
      <CompositeItem state={ state }>Item A2</CompositeItem>
      <CompositeItem state={ state }>Item A3</CompositeItem>
    </CompositeGroup>
    <CompositeGroup role="row" state={ state }>
      <CompositeItem state={ state }>Item B1</CompositeItem>
      <CompositeItem state={ state }>Item B2</CompositeItem>
      <CompositeItem state={ state }>Item B3</CompositeItem>
    </CompositeGroup>
    <CompositeGroup role="row" state={ state }>
      <CompositeItem state={ state }>Item C1</CompositeItem>
      <CompositeItem state={ state }>Item C2</CompositeItem>
      <CompositeItem state={ state }>Item C3</CompositeItem>
    </CompositeGroup>
  </Composite>
);
```

...which is a much better reflection of what a developer is going to be doing.

So I suppose where you're seeing _"over-engineered"_ I'm seeing _"trying to be helpful for consumers"_ 🤷 I'm quite happy to add comments to the `transform` function (should have done that anyway), but I will always err on some additional developer complexity if it makes the docs easier to consume, particularly as `story` files are really only intended for Storybook.

Reply:
I see where you're coming from and it is kind of cool, but I think the point is that we need to remember that this is a niche _and_ deprecated component. We foresee very low levels of docs traffic and usage, thus Marco's initial emphasis on keeping this a minimal effort undertaking.

The only downsides to this transformer is the initial engineering effort and the future maintenance cost, so since it's already written and there will be no future maintenance, I think we can keep it as is.

Reply:
Added some comments to make it clearer.

For whatever it's worth, it took less time to put this `transform` function together than it would have done to write out the code block for each example. So point definitely acknowledged about the minimal effort undertaking, but of all the things that could have been done differently, this probably isn't top of the list for revisiting.